### PR TITLE
Fixes #2294. Reports errors in initialization

### LINKED ISSUE #2294: Error messages in initialization are not displayed.
#### Summary:
In v2.15.0, error messages from initialization are not being displayed.

#### Description:
This was introduced with the refactor. The fix is quick, but I'll add some tests so the output doesn't disappear in the future.

#### Reproducible Steps:

Stan program:
```
parameters {
  real<lower = 0, upper = 1> theta[1];
}
model {
  theta[2] ~ normal(0, 1);
}
```

I'm compiling using CmdStan.

#### Current Output:
```method = sample (Default)
  sample
    num_samples = 1000 (Default)
    num_warmup = 1000 (Default)
    save_warmup = 0 (Default)
    thin = 1 (Default)
    adapt
      engaged = 1 (Default)
      gamma = 0.050000000000000003 (Default)
      delta = 0.80000000000000004 (Default)
      kappa = 0.75 (Default)
      t0 = 10 (Default)
      init_buffer = 75 (Default)
      term_buffer = 50 (Default)
      window = 25 (Default)
    algorithm = hmc (Default)
      hmc
        engine = nuts (Default)
          nuts
            max_depth = 10 (Default)
        metric = diag_e (Default)
        stepsize = 1 (Default)
        stepsize_jitter = 0 (Default)
id = 0 (Default)
data
  file = examples/bernoulli/bernoulli.data.R
init = 2 (Default)
random
  seed = 3072720890
output
  file = output.csv (Default)
  diagnostic_file =  (Default)
  refresh = 100 (Default)


Rejecting initial value:
  Error evaluating the log probability at the initial value.

Rejecting initial value:
  Error evaluating the log probability at the initial value.

Rejecting initial value:
  Error evaluating the log probability at the initial value.
...
```

#### Expected Output:

```
...
Rejecting initial value:
  Error evaluating the log probability at the initial value.

Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:
Exception thrown at line 5: []: accessing element out of range. index 2 out of range; expecting index to be between 1 and 1; index position = 1theta
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.
...
```

#### Current Version:
v2.15.0

### PR DESCRIPTION
**Pull request by syclik**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

Exceptions thrown in initialization are now reported.

#### How to Verify

Run 
```
./runTests.py src/test/unit/services/util/initialize_test.cpp 
```
or create a Stan program with a `reject()` statement. The tagged version won't output the error message.

#### Side Effects

None.

#### Documentation

None.

#### Reviewer Suggestions

Anyone.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Stan Group

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by syclik** [comment]
The message is coming from the model used to test the behavior. 

See line 198 of the test file. (I didn't change it in this pull request.)

**Comment by syclik** [comment]
@seantalts, any idea why the upstream tests are failing?

**Comment by seantalts** [comment]
Jenkins, retest this please.

It ran while I was fixing the upstream tests to work with forks.

File: src/test/unit/services/util/initialize_test.cpp [hunk 1]
```diff
@@ -256,15 +257,53 @@ namespace test {
...
+  std::vector<double> params;
+  test::mock_throwing_model throwing_model;
+
   double init_radius = 2;
   bool print_timing = false;
-  EXPECT_THROW(params
-               = stan::services::util::initialize(throwing_model, empty_context, rng,
-                                                  init_radius, print_timing,
-                                                  message, init),
+  EXPECT_THROW(params = stan::services::util::initialize(throwing_model, init_context, rng,
```
Comment (seantalts) [inline src/test/unit/services/util/initialize_test.cpp]: Out of curiosity, why do you have to assign params to the result here? Is it so the compiler doesn't optimize the call away?

Comment (syclik) [inline src/test/unit/services/util/initialize_test.cpp]: I don't need to. It was from copy and pasting a different test. I'll update the tests.

**Comment by syclik** [comment]
@seantalts, I've updated the tests. Feel free to merge.

**Pull request merged**